gh-103509: PEP 697 -- Limited C API for Extending Opaque Types (#103511)
python/cpython: Doc/c-api/object.rst

   returns ``NULL`` if the object cannot be iterated.

   .. versionadded:: 3.10

> .. c:function:: PyObject* PyObject_GetTypeData(PyObject *o, PyTypeObject *cls)

Reviewer: ```suggestion
.. c:function:: void *PyObject_GetTypeData(PyObject *o, PyTypeObject *cls)
```

>    Get a pointer to subclass-specific data reserved for *cls*.

Reviewer 2: For a new extension developer it may seem like getting class-specific metadata. Probably, something like this (but less clumsy) would be better:

```suggestion
   Get a pointer to a part of instance memory space of *o* reserved for *cls* level of an inheritance chain.
```

Author: Hm. Would this work?
```suggestion
   Get a pointer to the memory space of *o* reserved for data specific to subclass *cls*.
```

Reviewer 2: > Hm. Would this work?

Looks good.

>    The object *o* **must** be an instance of *cls*, and *cls* must have been

Reviewer: Observation: only one of the _must_'s are emphasised.

Author: ```suggestion
   The object *o* must be an instance of *cls*, and *cls* must have been
```

Reviewer: Yeah, that would work.

   created using negative :c:member:`PyType_Spec.basicsize`.
   Python does not check this.

>    On error, set an exception and return ``NULL``.

Reviewer: Again, no exception is set AFAICS. There is no error checking apart from a debug assert.

Author: The current implementation might not raise exceptions, but they may be added in the future. Users must implement error handling.

   .. versionadded:: 3.12

.. c:function:: Py_ssize_t PyType_GetTypeDataSize(PyTypeObject *cls)

>    Return the size of the memory reserved for *cls*, i.e. the size of the

Reviewer 2: The same suggestion as above.

```suggestion
   Return the size of the instance memory space reserved for *cls*, i.e. the size of the
```

   memory :c:func:`PyObject_GetTypeData` returns.

   This may be larger than requested using :c:member:`-PyType_Spec.basicsize <PyType_Spec.basicsize>`;
   it is safe to use this larger size (e.g. with :c:func:`!memset`).

   The type *cls* **must** have been created using
   negative :c:member:`PyType_Spec.basicsize`.
   Python does not check this.

>    On error, set an exception and return a negative value.

Reviewer: AFAICS, this function always succeeds; an exception is never set.

Return value is always >= 0.

Author: The current implementation might not raise exceptions, but they may be added in the future. Users must implement error handling.

Reviewer: That seems like a weird decision to me; if the current implementation does not raise exceptions, surely the current docs must reflect that fact. If the implementation changes, we change the docs. No?

Reviewer: Expanding: the runtime will segfault if the user does not _set an exception_ if any of these APIs indicate error. It is an error to return `NULL` from a function whose return value is `PyObject *` without an exception set.

Reviewer: ```suggestion
   This function always succeeds.
```

Reviewer: Wait, I think I see where you are heading. Yeah, you're good to go. (I just needed another cup of coffee to get this straight.)

   .. versionadded:: 3.12

> .. c:function:: PyObject* PyObject_GetItemData(PyObject *o)

Reviewer: ```suggestion
.. c:function:: void *PyObject_GetItemData(PyObject *o)
```

>    Get a pointer to per-item data for a class with

Reviewer 2: The same suggestion as above.

```suggestion
   Get a pointer to variable-length instance data not belonging to any class (level) in the inheritance chain.
```

Author: This PR adds the concept of data being reserved for particular (sub)classes/levels in the chain. 
`PyObject_GetItemData` should work with existing classes, where no data is tied to a subclass.
Also, variable-sized data *never* belongs to a specific subclass, even with this PR.

So IMO the original is better. It can probably still be improved though.

   :const:`Py_TPFLAGS_ITEMS_AT_END`.

   On error, set an exception and return ``NULL``.
   :py:exc:`TypeError` is raised if *o* does not have

>    :const:`Py_TPFLAGS_ITEMS_AT_END` set.

Reviewer: AFAICS, this function always succeeds; an exception is never set.

`TypeError` is not raised if `Py_TPFLAGS_ITEMS_AT_END` is not set; there's only an debug assertion.

Author: Good catch, thanks!
